actors/evm/src/interpreter/precompiles.rs

```diff
@@ -57,141 +82,181 @@ impl<BS: Blockstore, RT: Runtime<BS>> Precompiles<BS, RT> {
+
+    if base_len == 0 && mod_len == 0 {
         return Ok(Vec::new());
     }
 
-    // TODO bounds checking
-    let mut start = 96;
-    let base = BigUint::from_bytes_be(&input[start..start + base_len]);
-    start += base_len;
-    let exponent = BigUint::from_bytes_be(&input[start..start + exponent_len]);
-    start += exponent_len;
-    let modulus = BigUint::from_bytes_be(&input[start..start + mod_len]);
+    let base = read_bigint(input, 0, base_len);
+    let exponent = read_bigint(input, base_len, exponent_len);
+    let modulus = read_bigint(input, base_len + exponent_len, mod_len);
 
-    let mut output = if modulus.is_zero() || modulus.is_one() {
-        BigUint::zero().to_bytes_be()
-    } else {
-        base.modpow(&exponent, &modulus).to_bytes_be()
-    };
+    if modulus.is_zero() || modulus.is_one() {
+        // mod 0 is undefined: 0, base mod 1 is always 0
+        return Ok(vec![0; mod_len]);
+    }
+
+    let mut output = base.modpow(&exponent, &modulus).to_bytes_be();
 
     if output.len() < mod_len {
         let mut ret = Vec::with_capacity(mod_len);
-        ret.extend(core::iter::repeat(0).take(mod_len - output.len()));
+        ret.extend(std::iter::repeat(0).take(mod_len - output.len())); // left padding
         ret.extend_from_slice(&output);
         output = ret;
     }
 
     Ok(output)
 }
 
-/// converts 2 byte arrays (U256) into a point on a field
-/// exits with OutOfGas for any failed operation
-fn uint_to_point(x: U256, y: U256) -> Result<G1, PrecompileError> {
-    let x = Fq::from_u256(x.0.into()).map_err(|_| PrecompileError::EcErr)?;
-    let y = Fq::from_u256(y.0.into()).map_err(|_| PrecompileError::EcErr)?;
+/// reads a point from 2, 32 byte coordinates with right padding
+/// exits with EcErr for any failed operation
+/// panics if input.len() < 64
+fn curve_point(input: &[u8]) -> Result<G1, PrecompileError> {
+    let x = Fq::from_u256(U256::from_big_endian(&input[..32]).into())?;
+    let y = Fq::from_u256(U256::from_big_endian(&input[32..64]).into())?;
 
-    Ok(if x.is_zero() && y.is_zero() {
-        G1::zero()
-    } else {
-        AffineG1::new(x, y).map_err(|_| PrecompileError::EcErr)?.into()
-    })
+    Ok(if x.is_zero() && y.is_zero() { G1::zero() } else { AffineG1::new(x, y)?.into() })
 }
 
-/// add 2 points together on `alt_bn128`
-fn ec_add<RT: Primitives>(_: &RT, input: &[u8]) -> PrecompileResult {
-    let x1 = read_u256_infalliable(input, 0);
-    let y1 = read_u256_infalliable(input, 32);
-    let point1 = uint_to_point(x1, y1)?;
-
-    let x2 = read_u256_infalliable(input, 64);
-    let y2 = read_u256_infalliable(input, 96);
-    let point2 = uint_to_point(x2, y2)?;
+fn curve_to_vec(curve: G1) -> Vec<u8> {
+    AffineG1::from_jacobian(curve)
+        .map(|product| {
+            let mut output = vec![0; 64];
+            product.x().to_big_endian(&mut output[..32]).unwrap();
+            product.y().to_big_endian(&mut output[32..]).unwrap();
```

Comment:
can these panic?

Reply:
no .x() is 32 bytes and output vec is 64 bytes